### ISSUE TITLE
Support Pipeline Operator

### ISSUE THREAD

**Issue by k15a** [facebook/flow#5443]
The pipeline `|>` operator is a proposal which is currently in stage 1. It would be great if flow could support this proposal.

I could imagine that this is not too hard because it's basically only syntactic sugar for function calls IIRC.

https://github.com/tc39/proposal-pipeline-operator

**Comment by Hypnosphi** [comment]
I would like to help implementing that. Where should I start?

**Comment by gajus** [comment]
It has landed in Firefox under `--enable-pipeline-operator` compile flag.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Pipeline_operator

**Comment by mAAdhaTTah** [comment]
I've commented this elsewhere, but I strongly suggest avoiding implementing the pipeline operator until we've at least selected which of the 3 proposals we're going with. It's been in Firefox for some time behind a flag, but my current understanding is the FF implement doesn't currently conform with any of the 3 proposals' current semantics.

**Comment by sartaj** [comment]
@mAAdhaTTah It seems [Proposal 0](https://github.com/tc39/proposal-pipeline-operator/wiki#proposal0-original-minimal-proposal) fits specs for all other proposals. Couldn't flow at least implement this minimal behavior to start? This would benefit anyone who would love to use piped unary functions with typed data structures.

**Comment by Hypnosphi** [comment]
Flow is still 0.x so I personally wouldn't mind if this would be supported in some form now and then changed later

**Comment by johnsonw** [comment]
I would really like to see this move forward. After spending quite some time in F# I find myself using the pipeline operator routinely. https://babeljs.io/docs/en/babel-plugin-proposal-pipeline-operator shows that we can use the "minimal" proposal at the moment. Is there any way to suppress the following error:

```
 21:   const c = xs |> fn1 |> fn2 |> fn3;
                            ^ Unexpected token >
```
// $FlowFixMe doesn't seem to ignore the error.

**Comment by pnevyk** [comment]
> // $FlowFixMe doesn't seem to ignore the error.

That's because $FlowFixMe is for suppressing type errors. However, this is "syntax error" for flow parser so it does not get into the type checking phase at all.

**Comment by bsmith-cycorp** [comment]
Please please please fix this. The basic spec (1-arity functions) seems very unlikely to change and relatively easy to parse, and it's pretty much all that people are asking for.

**Comment by mAAdhaTTah** [comment]
> The "minimal" spec (1-arity functions) seems very unlikely to change

Do not assume this. There are 3 proposals, and the "minimal" spec is unlikely to be accepted as-is.

**Comment by Hypnosphi** [comment]
> the "minimal" spec is unlikely to be accepted as-is

Its benefit is not that it's the one that will be accepted most likely, but the fact that other proposal are supersets of the minimal one. So if you support the minimal proposal now, there will be no breaking changes required to support whichever of the advanced proposal wins later

**Comment by mAAdhaTTah** [comment]
> the fact that other proposals are supersets of the minimal one.

They are not. This is valid in minimal:

```
x |> f(y)
```

but not currently valid in smart (requires a placeholder token).

**Comment by gajus** [comment]
It would be nice to support this that Babel added support:

* https://github.com/babel/babel/blob/master/CHANGELOG.md#v750-2019-07-04
* https://babeljs.io/docs/en/babel-plugin-proposal-pipeline-operator

**Comment by goodmind** [comment]
@gajus which one? F# or smart?

**Comment by gajus** [comment]
> F# Pipelines

Based on the conversations I have been seeing, it is extremely unlikely that smart is going to be implemented. F# Pipelines is the way forward.

**Comment by goodmind** [comment]
Can't find what stage it is on

**Comment by noppa** [comment]
There are _three_ different proposals being considered, the two mentioned in the proposal readme and [Elixir-style pipelines](https://github.com/tc39/proposal-pipeline-operator/issues/143). It would be pretty amazing if Flow could support all three with a configuration option to select which one (if any) to use in a project. That way people could experiment with the different proposals and give feedback on them. Although of course adding support for all three is a lot more work than for just one.

**Comment by goodmind** [comment]
@mAAdhaTTah can you confirm that smart pipelines is unlikely to be implemented?

**Comment by goodmind** [comment]
@noppa 
![this is getting out of hand](https://media1.tenor.com/images/32097ca25431ec205fafc365d00c4f9c/tenor.gif?itemid=11488732)
`s/two/three`

**Comment by goodmind** [comment]
I might give a try at implementing F# pipeline operator parsing in Flow (without typechecking)

**Comment by mAAdhaTTah** [comment]
No, I can not confirm that. Until a final decision is made, all options are on the table.

**Comment by chicoxyzzy** [comment]
The final decision on which proposal to choose most likely won't be made until at least October TC39 meeting or even later

**Comment by goodmind** [comment]
@noppa I can't find Elixir proposal itself, only mentions about it

**Comment by noppa** [comment]
@goodmind I don't think there's anything more at the moment than the issue tc39/proposal-pipeline-operator#148, with some pretty good arguments for it ([comment](https://github.com/tc39/proposal-pipeline-operator/issues/143#issuecomment-492019845)) and against ([comment](https://github.com/tc39/proposal-pipeline-operator/issues/143#issuecomment-449612782), [comment](https://github.com/tc39/proposal-pipeline-operator/issues/143#issuecomment-459506746)).

I don't think there's actually a Babel plugin for it yet and I'm not sure if the specifics of it are defined well enough to actually implement yet... Sorry if I misled you by calling it a _proposal_ when at the moment it seems to be more like a _suggestion_.

If you are doing the F#-style proposal, btw, there's _also_ [partial application](https://github.com/tc39/proposal-partial-application), which is a separate proposal but would go well with the F# style pipelines.

**Comment by mAAdhaTTah** [comment]
Yeah, there currently aren't any active plans to implement Elixir-style pipelines in Babel, nor are there specs for it, although Daniel (the proposal champion) has been pushing a "try everything" philosophy, so we may end up doing it.

**Comment by goodmind** [comment]
@gajus Well, I guess if you want to try F# pipelines you can build my branch

**Comment by gkz** [comment]
8 years later, not much progress made on the proposal

**Comment by gajus** [comment]
I remember when this issue was raised and I was planning to start a new project, I thought to myself 'Pipeline operator will help a ton here. Better just wait until it lands' 😂

**Comment by js-choi** [comment]
* It’s totally reasonable to wait on implementing this in Flow until the pipe operator progresses further, i.e., to Stage 3.
* Re: not much progress—I gave a general proposal update in https://github.com/tc39/proposal-pipeline-operator/issues/232#issuecomment-2784879225.
  * TL’DR: The Committee has become progressively chillier/stingier toward new syntax, and because there are interactions with the syntaxes of certain other proposals.
  * So the pipe champions been focusing on lower-hanging non-syntax API proposals for now.
  * The proposal is not dead. We’re just biding our time to see how composite literals (formerly records/tuples) and JS0/JSSugar shake out.
* Hopefully we’ll be able to progress the proposal again in a few years, only at which it would be appropriate for Flow to implement support.
